packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,374 @@
+ * Populates {@link identifierCounts} by the number of times each type parameter
+ * appears in the given type by recursively descending through type references.
+ * This is essentially a limited subset of the scope manager, but for types.
+ */
+function collectTypeParameterUsageCounts(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  identifierCounts: Map<ts.Identifier, number>,
+): void {
+  const visitedSymbolLists = new Set<ts.Symbol[]>();
+  const type = checker.getTypeAtLocation(node);
+  const typeCounts = new Map<ts.Type, number>();
+  const visitedConstraints = new Set<ts.TypeNode>();
+  let functionLikeType = false;
+  let visitedDefault = false;
+
+  if (
+    ts.isCallSignatureDeclaration(node) ||
+    ts.isConstructorDeclaration(node)
+  ) {
+    functionLikeType = true;
+    visitSignature(checker.getSignatureFromDeclaration(node));
+  }
+
+  if (!functionLikeType) {
+    visitType(type, false);
+  }
+
+  function visitType(type: ts.Type | undefined, asRepeatedType: boolean): void {
+    // Seeing the same type > (threshold=3 ** 2) times indicates a likely
+    // recursive type, like `type T = { [P in keyof T]: T }`.
+    // If it's not recursive, then heck, we've seen it enough times that any
+    // referenced types have been counted enough to qualify as used.
+    if (!type || incrementTypeCount(type) > 9) {
+      return;
+    }
+
+    // https://github.com/JoshuaKGoldberg/ts-api-utils/issues/382
+    if ((tsutils.isTypeParameter as (type: ts.Type) => boolean)(type)) {
+      const declaration = type.getSymbol()?.getDeclarations()?.[0] as
+        | ts.TypeParameterDeclaration
+        | undefined;
+
+      if (declaration) {
+        incrementIdentifierCount(declaration.name, asRepeatedType);
+
+        // Visiting the type of a constrained type parameter will recurse into
+        // the constraint. We avoid infinite loops by visiting each only once.
+        if (
+          declaration.constraint &&
+          !visitedConstraints.has(declaration.constraint)
+        ) {
+          visitedConstraints.add(declaration.constraint);
+          visitType(checker.getTypeAtLocation(declaration.constraint), false);
+        }
+
+        if (declaration.default && !visitedDefault) {
+          visitedDefault = true;
+          visitType(checker.getTypeAtLocation(declaration.default), false);
+        }
+      }
+    }
+
+    // Intersections and unions like `0 | 1`
+    else if (tsutils.isUnionOrIntersectionType(type)) {
+      visitTypesList(type.types, asRepeatedType);
+    }
+
+    // Index access types like `T[K]`
+    else if (tsutils.isIndexedAccessType(type)) {
+      visitType(type.objectType, asRepeatedType);
+      visitType(type.indexType, asRepeatedType);
+    }
+
+    // Tuple types like `[K, V]`
+    // Generic type references like `Map<K, V>`
+    else if (tsutils.isTupleType(type) || tsutils.isTypeReference(type)) {
+      for (const typeArgument of type.typeArguments ?? []) {
+        visitType(typeArgument, true);
```

Comment:
I understand the idea behind this for type aliases (see my long comment above) but I don't understand it for tuples. In fact, I think it might be wrong! This is considered valid code by this rule but it should not be:

```ts
function takeTuple<T>(x: [T]): void;
```

Reply:
Wouldn't array methods like `map` also count as references to `T`, though?

Reply:
Good point, I guess they would.

I do think it would be interesting to explore adding a list of "singular types" to this rule in the future. `Array<T>` does, in theory, define a complex set of relationships between methods involving `T`. But in practice, you probably just want to pass in a JS array. Same for `Set<T>`, `Map<K, V>` and I'm sure other types, too.